mesonbuild/interpreter.py

```diff
@@ -2413,15 +2413,15 @@ def func_dependency(self, node, args, kwargs):
                 exception = DependencyException("fallback for %s not found" % name)
 
             # Search inside the projects list
-            if not dep or not dep.found():
+            if not dep.found():
                 if 'fallback' in kwargs:
                     fallback_dep = self.dependency_fallback(name, kwargs)
                     if fallback_dep:
                         # Never add fallback deps to self.coredata.deps since we
                         # cannot cache them. They must always be evaluated else
                         # we won't actually read all the build files.
                         return fallback_dep
-                if not dep:
+                if required:
                     assert(exception is not None)
                     raise exception
```

Comment:
This code overall became extremely messy. We throw exception created in else branch of previous statement. We should refactor it. I will do it in next PR or you can do it now, whatever you prefer. But this has to be done.

Reply:
This section of code is extremely messy, yes, and I was hesitant to change it so soon before release because lots of regressions have happened in the past by people trying to change it without understanding all the edge cases.

I would recommend reorganising it after 0.46.

Reply:
Sounds good.